yt_dlp/extractor/niconico.py

```diff
@@ -33,13 +32,72 @@
     urlencode_postdata,
     urljoin,
 )
+from ..utils.traversal import find_element, traverse_obj
 
 
-class NiconicoIE(InfoExtractor):
+class NiconicoBaseIE(InfoExtractor):
+    _GEO_BYPASS = False
+    _GEO_COUNTRIES = ['JP']
+    _LOGIN_BASE = 'https://account.nicovideo.jp'
+    _NETRC_MACHINE = 'niconico'
+
+    def raise_login_error(self, error, default, expected=True):
+        raise ExtractorError(f'Unable to login: {error or default}', expected=expected)
+
+    def _perform_login(self, username, password):
+        if self._get_cookies('https://www.nicovideo.jp').get('user_session'):
+            return
+
+        self._request_webpage(
+            f'{self._LOGIN_BASE}/login', None, 'Requesting session cookies')
+        webpage = self._download_webpage(
+            f'{self._LOGIN_BASE}/login/redirector', None,
+            'Logging in', headers={
+                'Content-Type': 'application/x-www-form-urlencoded',
+                'Referer': f'{self._LOGIN_BASE}/login',
+            }, data=urlencode_postdata({
+                'mail_tel': username,
+                'password': password,
+            }),
+        )
+
+        if traverse_obj(self._search_json(
+            r'NicoGoogleTagManagerDataLayer\s*=\s*\[',
+            webpage, 'data layer', None, fatal=False,
+        ), ('user', 'login_status', {lambda x: x == 'login'})):
+            return
```

Comment:
IMO it's okay to break this up to make it more readable:
```suggestion
        data_layer = self._search_json(
            r'NicoGoogleTagManagerDataLayer\s*=\s*\[', webpage, 'data layer', None, fatal=False)
        if traverse_obj(data_layer, ('user', 'login_status')) == 'login':
            return
```

&nbsp;

Or, could we just do this?

```suggestion
        if self._get_cookies('https://www.nicovideo.jp').get('user_session'):
            return
```

Reply:
It's possible if checking `user_session` twice (at the start of `_perform_login` and after login) is allowed
something like this?
```python
@property
def is_logged_in(self):
    return bool(self._get_cookies('https://www.nicovideo.jp').get('user_session'))
```

Reply:
sure